[RESOLVE]#858 Rework passive assignment logic

DannyDelic opened the pull request:
- Renamed variable to a more appropriate name
- Now finds name by checking the attached Passive.cs file in a champion folder
- Parses the *.cs-file as text and uses a regex match, to find the passiveName
- Should be somewhat resilliant towards similar-sounding comments, as it checks for the following interface in the pattern

DannyDelic committed:
#858 Rework passive assignment logic:

- Renamed variable to a more appropriate name
- Now finds name by checking the attached Passive.cs file in a champion folder
- Parses the *.cs-file as text and uses a regex match, to find the passiveName
- Should be somewhat resilliant towards similar-sounding comments, as it checks for the following interface in the pattern

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/LeagueSandbox/GameServer?pullRequest=860) <br/>All committers have signed the CLA.

abosma commented on GameServerLib/Content/CharData.cs:
```diff
@@ -10,8 +11,22 @@ namespace LeagueSandbox.GameServer.Content
+        //TODO: Extend into handling several passives, when we decide on a format for that case.
+        public static string GetPassiveAbilityNameFromScriptFile(string champName)
+        {
+            var inputPassiveFile = File.ReadAllText($"..\\..\\..\\Content\\LeagueSandbox-Scripts\\Champions\\{champName}\\Passive.cs");
```
You would rather not hard-code this to LeagueSandbox-Scripts.

DannyDelic replied:
Is there any environment-variable I can point at in this case then?

abosma replied:
Game has the config file for the environment variables like content location, which I believe CharData contains.

lzardy replied:
Check Package.cs as well if you haven't already.

DannyDelic replied:
There is an attached _contentManager in the CharData class, but the variables for the PackagePath and Package is empty - so I have been forced to alter the ICharData interface to get the desired strings, which you can see in my next commit.

abosma replied:
Using _contentManager.GetAllLoadedPackages() gives me all packages loaded together with their names and paths, here's and example picture: [Code example](https://i.imgur.com/vAudcSW.png)

DannyDelic replied:
Didn't load the packages, thanks :)

DannyDelic replied:
Should be resolved in the latest commit now.

DannyDelic committed:
Rework passive assignment logic: Formatting-fix

DannyDelic committed:
#858 Rework passive assignment logic

- Changed hardcoded path in the parser-method to contain environment-variables.

DannyDelic committed:
#858 Rework passive assignment logic

- Now uses the more generic IPackage.

DannyDelic commented:
> You should use an IPackage list instead of Package since there is probably going to be an update for different kinds of packages.

Noted and implemented.

danil179 commented on GameServerLib/Content/CharData.cs:
```diff
@@ -10,8 +13,32 @@ namespace LeagueSandbox.GameServer.Content
+                    RegexOptions options = RegexOptions.Multiline;
+
+                    var passiveName = Regex.Match(inputPassiveFile, pattern, options).Groups["passiveName"].Value;
+
```
Remove blank lines.

danil179 commented on GameServerLib/Content/CharData.cs:
```diff
@@ -10,8 +13,32 @@ namespace LeagueSandbox.GameServer.Content
+                {
+                    var inputPassiveFile = File.ReadAllText(path);
+
+                    string pattern = @"class (?<passiveName>\w+) : IGameScript";
```
We need to discuss whether we want to use a way like that (inferring names from classes. Maybe we want to do some reflection magic to have a better handling)

DannyDelic replied:
I don't have much experience using the Reflections library, could you explain how you would want to use it for this use-case`?

danil179 replied:
See this https://github.com/LeagueSandbox/GameServer/blob/indev/GameServerLib/Scripting/CSharp/CSharpScriptEngine.cs (this class uses a lot of reflection).

It should be something like:
* Get the file
* Create some assembly object from it
* Get class name
* Infer passive name from it

danil179 replied:
Let's delay it for another PR as compiling in real time only to find the passive name is not efficient at all. (We need to add functions to the `CSharpScriptEngine` to allow that)

danil179 reviewed (changes requested):
Fix the style issues and we will discuss this way of deducing the passive names.

DannyDelic committed:
#858 Rework passive assignment logic

- Fix style issues

danil179 reviewed (approved):
Approved. Issue will be added regrading the problematic regex (I can see how one can misuse this to break abilities/ crash the game).
